src/helics/shared_api_library/helicsData.cpp

```diff
@@ -234,13 +303,240 @@ double helicsDataBufferToDouble(HelicsDataBuffer data)
+{
+    auto* ptr = getBuffer(data);
+    if (ptr == nullptr) {
+        return HELICS_FALSE;
+    }
+    char val;
+    helics::valueExtract(helics::data_view(*ptr), helics::detail::detectType(ptr->data()), val);
     return val;
 }
+
+HelicsTime helicsDataBufferToTime(HelicsDataBuffer data)
+{
+    auto* ptr = getBuffer(data);
+    if (ptr == nullptr) {
+        return HELICS_TIME_INVALID;
+    }
+    helics::Time val;
+    helics::valueExtract(helics::data_view(*ptr), helics::detail::detectType(ptr->data()), val);
+    return static_cast<HelicsTime>(val);
+}
+
+HelicsComplex helicsDataBufferToComplexObject(HelicsDataBuffer data)
+{
+    auto* ptr = getBuffer(data);
+    HelicsComplex val{HELICS_TIME_INVALID, HELICS_TIME_INVALID};
+    if (ptr != nullptr) {
+        std::complex<double> v;
+        helics::valueExtract(helics::data_view(*ptr), helics::detail::detectType(ptr->data()), v);
+        val.real = v.real();
+        val.imag = v.imag();
+    }
+    return val;
+}
+
+void helicsDataBufferToComplex(HelicsDataBuffer data, double* real, double* imag)
+{
+    auto* ptr = getBuffer(data);
+    if (ptr != nullptr) {
+        std::complex<double> v;
+        helics::valueExtract(helics::data_view(*ptr), helics::detail::detectType(ptr->data()), v);
+        if (real != nullptr) {
+            *real = v.real();
+        }
+        if (imag != nullptr) {
+            *imag = v.imag();
+        }
+        return;
+    }
+    if (real != nullptr) {
+        *real = HELICS_TIME_INVALID;
+    }
+    if (imag != nullptr) {
+        *imag = HELICS_TIME_INVALID;
+    }
+}
+
+int helicsDataBufferStringSize(HelicsDataBuffer data)
+{
+    auto* ptr = getBuffer(data);
+    if (ptr == nullptr) {
+        return 0;
+    }
+    switch (helics::detail::detectType(ptr->data())) {
+        case helics::DataType::HELICS_STRING:
+            return static_cast<int>(helics::detail::getDataSize(ptr->data()));
+        default: {
+            std::string v;
+            helics::valueExtract(helics::data_view(*ptr), helics::detail::detectType(ptr->data()), v);
+            return static_cast<int>(v.size());
+        } break;
+    }
+}
+
+void helicsDataBufferToString(HelicsDataBuffer data, char* outputString, int maxStringLen, int* actualLength)
+{
+    if ((outputString == nullptr) || (maxStringLen <= 0)) {
+        if (actualLength != nullptr) {
+            *actualLength = 0;
+        }
+        return;
+    }
+    auto* ptr = getBuffer(data);
+    if (ptr == nullptr) {
+        if (actualLength != nullptr) {
+            *actualLength = 0;
+        }
+        return;
+    }
+    std::string v;
+    helics::valueExtract(helics::data_view(*ptr), helics::detail::detectType(ptr->data()), v);
+
+    auto length = (std::min)(static_cast<int>(v.size()), maxStringLen);
+    std::memcpy(outputString, v.data(), length);
+    if (actualLength != nullptr) {
+        *actualLength = length;
+    }
+    // add a null terminator if possible
```

Comment:
I think ending the string with a null terminator should be consistent regardless of buffer space -- probably safer with a null terminator assuming the data is guaranteed to not have null bytes, but if the helics data buffer being converted can contain any data bytes then we should never add a null terminator.

Reply:
I split this into two distinct functions one guarantees a null terminator, the other does not do anything with null terminators.    And the string size method was modified so the size returned includes space for a null terminator.